cache/cache_rep.h

```diff
@@ -0,0 +1,45 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+#include "table/block_based/block_based_table_reader.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+class CacheRep {
```

Comment:
[non-blocking][NEED_ATTENTION]
Question 1: what is the role of such a CacheRep class for class users other than WriteBufferManager(e.g, other dummy entries clients like BlockBasedTableBuilder)?

This question needs attention since (1) there might be some follow-up work on top of this working solution, depending on the answer (e.g, a trivial one would be renaming the class) (2)I need to write a class-level description for my new file.

I don't have a very clear answer to the question above due to that:
(1)The existing design of CacheRep did more than just a simple C-style struct (an aggregate class) since we have GetNextCacheKey() and my improved design naturally encapsulated some of its impl details by using private data members.
(2)The current CacheRep does less than fully encapsulating the impl details of dummy entry manipulation (e.g, inserting/releasing/boo-keeping) since dummy entry client (e.g, WriteBufferManager) still needs to interact with the underlying "cache_" or "cache_allocated_size_" for dummy entry manipulation. 

It is also a bit hard for such encapsulation since different dummy entry clients have slightly different needs for such a CacheRep. For example, BlockBasedTableBuilder might not need any bookkeeping on "cache_allocated_size_" at all. This also reflects the issue of not all the public data members in the interface are useful for all class users.

Therefore (1) and (2) lead to some sort of description of CacheRep being "an aggregate class for dummy entry manipulation but with encapsulation on dummy entry cache key generation and with part of the public interface not being applicable to some class users" .

I found the description above for CacheRep class a bit awkward - but may be just me feeling this way...